pkg/boot/kexec/kexec_linux_implemented.go

```diff
@@ -26,8 +34,38 @@ func FileLoad(kernel, ramfs *os.File, cmdline string) error {
 		flags |= unix.KEXEC_FILE_NO_INITRAMFS
 	}
 
+	if isGzip(kernel) {
+		g, err := gzip.NewReader(kernel)
+		if err != nil {
+			return fmt.Errorf("could read kernel gzip header: %v", err)
+		}
+		decompressed, err := ioutil.TempFile("", "kernel")
+		if err != nil {
+			return fmt.Errorf("could not create temp file: %v", err)
+		}
+		defer os.Remove(decompressed.Name())
+		if _, err := io.Copy(decompressed, g); err != nil {
+			return fmt.Errorf("could not decompress kernel: %v", err)
+		}
+		if err := g.Close(); err != nil {
+			return fmt.Errorf("could not close gunzip (bad checksum?): %v", err)
+		}
+		if err := decompressed.Sync(); err != nil {
+			return fmt.Errorf("could not sync decompressed kernel: %v", err)
+		}
+		kernel = decompressed
+	}
+
 	if err := unix.KexecFileLoad(int(kernel.Fd()), ramfsfd, cmdline, flags); err != nil {
 		return fmt.Errorf("sys_kexec(%d, %d, %s, %x) = %v", kernel.Fd(), ramfsfd, cmdline, flags, err)
 	}
 	return nil
 }
+
+// isGzip returns true if the file is gzip file format.
```

Comment:
amazed that this is not in the package ...

Reply:
You should see how complex the C version is: https://github.com/horms/kexec-tools/blob/master/kexec/arch/arm64/kexec-zImage-arm64.c